plugins/pixart-tp/fu-pixart-tp-section.c

```diff
@@ -0,0 +1,245 @@
+	if (self->flags != FU_PIXART_TP_FIRMWARE_FLAG_NONE) {
+		g_autofree gchar *str = fu_pixart_tp_firmware_flags_to_string(self->flags);
+		fu_xmlb_builder_insert_kv(bn, "flags", str);
+	}
+	fu_xmlb_builder_insert_kx(bn, "target_flash_start", self->target_flash_start);
+	fu_xmlb_builder_insert_kx(bn, "section_crc", self->section_crc);
+	if (self->reserved != NULL) {
+		g_autofree gchar *str = fu_byte_array_to_string(self->reserved);
+		fu_xmlb_builder_insert_kv(bn, "reserved", str);
+	}
+}
+
+static gboolean
+fu_pixart_tp_section_build(FuFirmware *firmware, XbNode *n, GError **error)
+{
+	FuPixartTpSection *self = FU_PIXART_TP_SECTION(firmware);
+	const gchar *tmp;
+	guint64 tmp64 = 0;
+
+	/* simple properties */
+	tmp = xb_node_query_text(n, "update_type", NULL);
+	if (tmp != NULL)
+		self->update_type = fu_pixart_tp_update_type_from_string(tmp);
+	tmp = xb_node_query_text(n, "flags", NULL);
+	if (tmp != NULL)
+		self->flags = fu_pixart_tp_firmware_flags_from_string(tmp);
+	tmp = xb_node_query_text(n, "target_flash_start", NULL);
+	if (tmp != NULL) {
+		if (!fu_strtoull(tmp, &tmp64, 0, G_MAXUINT32, FU_INTEGER_BASE_AUTO, error))
+			return FALSE;
+		self->target_flash_start = (guint32)tmp64;
+	}
+	tmp = xb_node_query_text(n, "section_crc", NULL);
+	if (tmp != NULL) {
+		if (!fu_strtoull(tmp, &tmp64, 0, G_MAXUINT32, FU_INTEGER_BASE_AUTO, error))
+			return FALSE;
+		self->section_crc = (guint32)tmp64;
+	}
+	tmp = xb_node_query_text(n, "reserved", NULL);
+	if (tmp != NULL) {
+		if (self->reserved != NULL)
+			g_byte_array_unref(self->reserved);
+		self->reserved = fu_byte_array_from_string(tmp, error);
+		if (self->reserved == NULL)
+			return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_pixart_tp_section_parse(FuFirmware *firmware,
+			   GInputStream *stream,
+			   gsize offset,
+			   FuFirmwareParseFlags flags,
+			   GError **error)
+{
+	FuPixartTpSection *self = FU_PIXART_TP_SECTION(firmware);
+	const guint8 *reserved_src = NULL;
+	gsize reserved_len = 0;
+	guint32 section_length;
+	g_autofree gchar *extname = NULL;
+	g_autoptr(FuStructPixartTpFirmwareSectionHdr) st = NULL;
+
+	g_return_val_if_fail(FU_IS_PIXART_TP_SECTION(self), FALSE);
+	g_return_val_if_fail(G_IS_INPUT_STREAM(stream), FALSE);
+
+	st = fu_struct_pixart_tp_firmware_section_hdr_parse_stream(stream, offset, error);
+	if (st == NULL)
+		return FALSE;
+
+	/* core fields */
+	self->update_type = fu_struct_pixart_tp_firmware_section_hdr_get_update_type(st);
+	self->flags = fu_struct_pixart_tp_firmware_section_hdr_get_update_info(st);
+	self->target_flash_start =
+	    fu_struct_pixart_tp_firmware_section_hdr_get_target_flash_start(st);
+	fu_firmware_set_offset(
+	    firmware,
+	    fu_struct_pixart_tp_firmware_section_hdr_get_internal_file_start(st));
+	section_length = fu_struct_pixart_tp_firmware_section_hdr_get_section_length(st);
+	self->section_crc = fu_struct_pixart_tp_firmware_section_hdr_get_section_crc(st);
```

Comment:
@PixartHarris I've noticed that some sections have a CRC, and some don't (e.g. CRC is set as zero) -- should we check this CRC here? if so, is the CRC a standard CRC of the section _data_ or the section _header_?